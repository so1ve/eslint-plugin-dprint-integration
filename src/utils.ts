import { readFileSync } from "node:fs";
import { extname } from "node:path";

const TS_EXTS = [".ts", ".tsx", ".js", ".jsx", ".mjs", ".cjs", ".mts", ".cts"];
const JSON_EXTS = [".json", ".jsonc", ".json5"];
const MARKDOWN_EXTS = [".md", ".markdown", ".mdown", ".mkd"];
const MALVA_EXTS = [".css", ".scss", ".sass", ".less"];
const MARKUP_EXTS = [".html", ".vue", ".svelte", ".twig", ".jinja", ".jinja2"];

export function detectLanguage(filename: string) {
	const ext = extname(filename);
	if (TS_EXTS.includes(ext)) {
		return "typescript";
	} else if (ext === ".toml") {
		return "toml";
	} else if (JSON_EXTS.includes(ext)) {
		return "json";
	} else if (MARKDOWN_EXTS.includes(ext)) {
		return "markdown";
	} else if (filename === "Dockerfile") {
		return "dockerfile";
	} else if (MALVA_EXTS.includes(ext)) {
		return "malva";
	} else if (MARKUP_EXTS.includes(ext)) {
		return "markup";
	}
}

export const hasNewlineOnly = (str: string) =>
	["\r", "\n", "\r\n"].includes(str);

export function pick<T extends Record<string, any>, K extends keyof T>(
	obj: T,
	pickKeys: K[] | readonly K[],
) {
	const result = {} as Pick<T, K>;
	for (const key of pickKeys) {
		if (obj[key] !== undefined) {
			result[key] = obj[key];
		}
	}

	return result;
}

export function omit<T extends Record<PropertyKey, unknown>, K extends keyof T>(
	obj: T,
	omitKeys: K[] | readonly K[],
) {
	const result = {} as Omit<T, K>;
	const keys = Object.keys(obj).filter(
		(item) => !omitKeys.includes(item as K),
	) as Exclude<keyof T, K>[];
	for (const key of keys) {
		if (obj[key] !== undefined) {
			result[key] = obj[key];
		}
	}

	return result;
}

export const readJson = (filename: string) =>
	JSON.parse(readFileSync(filename, "utf8"));
