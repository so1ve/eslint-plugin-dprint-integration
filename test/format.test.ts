import { join } from "node:path";

import { ESLint } from "eslint";
import { describe, expect, it } from "vitest";

import { dir } from "./utils";

const dirname = dir();
const relative = (path: string) => join(dirname, path);

const eslint = new ESLint({
	baseConfig: {
		parserOptions: {
			ecmaVersion: "latest",
			ecmaFeatures: {
				jsx: true,
			},
			sourceType: "module",
		},
		plugins: ["dprint-integration"],
		rules: {
			"dprint-integration/dprint": [
				"error",
				{ useDprintJson: relative("__fixtures__/dprint.json") },
			],
		},
	},
	useEslintrc: false,
	ignore: false,
});

const runFixtures = (fixtures: string) =>
	eslint
		.lintFiles(relative(fixtures))
		.then((result) => result.flatMap((r) => r.messages));

describe("format", () => {
	it("fixtures", async () => {
		const result = await runFixtures("./__fixtures__/format/**");

		expect(result).toMatchSnapshot();
	});
});
