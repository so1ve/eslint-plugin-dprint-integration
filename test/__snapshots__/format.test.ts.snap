// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`format > fixtures 1`] = `
[
  {
    "column": 6,
    "endColumn": 15,
    "endLine": 1,
    "fix": {
      "range": [
        5,
        14,
      ],
      "text": "</div>
",
    },
    "line": 1,
    "message": "Replace \`···</div>\` with \`</div>⏎\`",
    "messageId": "replace",
    "nodeType": null,
    "ruleId": "dprint-integration/dprint",
    "severity": 2,
  },
  {
    "column": 11,
    "endColumn": 15,
    "endLine": 1,
    "fix": {
      "range": [
        10,
        14,
      ],
      "text": "'1';
",
    },
    "line": 1,
    "message": "Replace \`"1";\` with \`'1';⏎\`",
    "messageId": "replace",
    "nodeType": null,
    "ruleId": "dprint-integration/dprint",
    "severity": 2,
  },
  {
    "column": 10,
    "endColumn": 11,
    "endLine": 1,
    "fix": {
      "range": [
        9,
        10,
      ],
      "text": "",
    },
    "line": 1,
    "message": "Delete \`·\`",
    "messageId": "delete",
    "nodeType": null,
    "ruleId": "dprint-integration/dprint",
    "severity": 2,
  },
  {
    "column": 1,
    "endColumn": 8,
    "endLine": 2,
    "fix": {
      "range": [
        12,
        19,
      ],
      "text": "  <div",
    },
    "line": 2,
    "message": "Replace \`↹<div··\` with \`··<div\`",
    "messageId": "replace",
    "nodeType": null,
    "ruleId": "dprint-integration/dprint",
    "severity": 2,
  },
  {
    "column": 1,
    "endColumn": 13,
    "endLine": 3,
    "fix": {
      "range": [
        27,
        39,
      ],
      "text": "</template>;
",
    },
    "line": 3,
    "message": "Replace \`·</template>\` with \`</template>;⏎\`",
    "messageId": "replace",
    "nodeType": null,
    "ruleId": "dprint-integration/dprint",
    "severity": 2,
  },
]
`;
